lib/src/dartdoc_options.dart

```diff
@@ -120,20 +120,135 @@ class CategoryConfiguration {
 /// [ToolConfiguration] class.
 class ToolDefinition {
   /// A list containing the command and options to be run for this tool. The
-  /// first argument in the command is the tool executable. Must not be an empty
-  /// list, or be null.
+  /// first argument in the command is the tool executable, and will have its
+  /// path evaluated relative to the `dartdoc_options.yaml` location. Must not
+  /// be an empty list, or be null.
   final List<String> command;
 
+  /// A list containing the command and options to setup phase for this tool.
+  /// The first argument in the command is the tool executable, and will have
+  /// its path evaluated relative to the `dartdoc_options.yaml` location. May
+  /// be null or empty, in which case it will be ignored at setup time.
+  final List<String> setupCommand;
+
   /// A description of the defined tool. Must not be null.
   final String description;
 
-  ToolDefinition(this.command, this.description)
+  /// If set, then the setup command has been run once for this tool definition.
+  bool setupComplete = false;
+
+  /// Returns true if the given executable path has an extension recognized as a
+  /// Dart extension (e.g. '.dart' or '.snapshot').
+  static bool isDartExecutable(String executable) {
+    var extension = pathLib.extension(executable);
+    return extension == '.dart' || extension == '.snapshot';
+  }
+
+  /// Creates a ToolDefinition or subclass that is appropriate for the command
+  /// given.
+  factory ToolDefinition.fromCommand(
+      List<String> command, List<String> setupCommand, String description) {
+    assert(command != null);
+    assert(command.isNotEmpty);
+    assert(description != null);
+    if (isDartExecutable(command[0])) {
+      return DartToolDefinition(command, setupCommand, description);
+    } else {
+      return ToolDefinition(command, setupCommand, description);
+    }
+  }
+
+  ToolDefinition(this.command, this.setupCommand, this.description)
       : assert(command != null),
         assert(command.isNotEmpty),
         assert(description != null);
 
   @override
-  String toString() => '$runtimeType: "${command.join(' ')}" ($description)';
+  String toString() {
+    final String commandString =
+        '${this is DartToolDefinition ? '(Dart) ' : ''}"${command.join(' ')}"';
+    if (setupCommand == null) {
+      return '$runtimeType: $commandString ($description)';
+    } else {
+      return '$runtimeType: $commandString, with setup command '
+          '"${setupCommand.join(' ')}" ($description)';
+    }
+  }
+}
+
+/// A singleton that keeps track of cached snapshot files. The [dispose]
+/// function must be called before process exit to clean up snapshots in the
+/// cache.
+class SnapshotCache {
```

Comment:
maybe a todo for the future, but it would make sense to me for the SnapshotCache to also create snapshots

Reply:
Well, but the problem is that in order to do that it would need to be the thing that also runs the tools, since the way the snapshots are created is to do a run of a tool with different args. It just seemed better to have it manage the cache, and have it just return paths of snapshots to be created instead of moving all of the tool running logic into the cache.  I could see just moving the caching logic into the `ToolRunner` itself, though.